gui/src/app/Project/ProjectDataModel.ts

```diff
@@ -1,21 +1,69 @@
-import {
-  SamplingOpts,
-  defaultSamplingOpts,
-  isSamplingOpts,
-} from "../StanSampler/StanSampler";
-
 export enum ProjectKnownFiles {
   STANFILE = "stanFileContent",
   DATAFILE = "dataFileContent",
 }
 
+const baseObjectCheck = (x: any): boolean => {
+  return (x ?? false) && typeof x === "object";
+};
+
+export type SamplingOpts = {
+  num_chains: number;
+  num_warmup: number;
+  num_samples: number;
+  init_radius: number;
+  seed: number | undefined;
+};
+
+export const isSamplingOpts = (x: any): x is SamplingOpts => {
+  if (!baseObjectCheck(x)) return false;
+  if (typeof x.num_chains !== "number") return false;
+  if (typeof x.num_warmup !== "number") return false;
+  if (typeof x.num_samples !== "number") return false;
+  if (typeof x.init_radius !== "number") return false;
+  if (x.seed !== undefined && typeof x.seed !== "number") return false;
+  return true;
+};
+
+const validateSamplingOpts = (x: SamplingOpts): boolean => {
+  const naturalFields = [x.num_chains, x.num_samples, x.num_warmup];
+  const positiveFields = [x.num_chains, x.num_samples];
+  const nonnegativeFields = [x.num_warmup, x.init_radius];
+  if (naturalFields.some((f) => !Number.isInteger(f))) return false;
+  if (positiveFields.some((f) => f <= 0)) return false;
+  if (nonnegativeFields.some((f) => f < 0)) return false;
+  return true;
+};
+
+export const parseSamplingOpts = (x: string | undefined): SamplingOpts => {
```

Comment:
Should we support partial fields? My thought is if we ever expose additional arguments, old links could break if we're insisting on a full parse.

I know we don't currently, so it's fine to not do this now, just speculating

Reply:
Yeah, it's a good question. So far we appear to be insisting on it, but fortunately for now the only way to get one of these (other than to craft it manually, which, why) would be to save one, and as it currently stands that will give a full object.

...but going forward, yeah, we should probably drop in defaults for non-provided values.